File: windows/src/BugNarrator.Core/Export/TrackerExportPayloadBudget.cs
namespace BugNarrator.Core.Export;

/// <summary>
/// Caps for text sent to external trackers, mirroring the macOS
/// `TrackerExportPayloadBudget` (Sources/BugNarrator/Services/TrackerExportPayloadBudget.swift).
///
/// Only the constants and helpers the reproduction-step rendering needs are ported here. Windows
/// previously applied no caps at all, which is tolerable for the flat fields but not for a nested
/// list the model controls: without a cap a single issue could push an arbitrarily large body at
/// GitHub or Jira, where macOS truncates. The marker strings match macOS exactly so an artifact
/// truncated on either platform reads the same.
/// </summary>
public static class TrackerExportPayloadBudget
{
    public const int ReproductionStepLimit = 10;
    public const int ListEntryLimit = 500;
    public const int ScreenshotListLimit = 10;

    /// <summary>
    /// Jira Cloud rejects a <c>summary</c> over 255 characters or containing a newline; GitHub
    /// rejects an issue <c>title</c> over 256. Hard server limits, unlike the body budgets, which
    /// are self-imposed. Jira counts UTF-16 code units — the same unit <see cref="string.Length"/>
    /// uses — so the cap here is exact for Jira (#1200, macOS #1112).
    /// </summary>
    public const int JiraSummaryLimit = 255;
    public const int GitHubTitleLimit = 256;

    private const string TruncationMarker = " …[truncated by BugNarrator for tracker limits]";
    private const string OmissionNotice = "Additional items were omitted by BugNarrator to fit tracker limits.";

    public static string Truncated(string value, int maxCharacters)
    {
        var trimmed = (value ?? string.Empty).Trim();
        if (trimmed.Length <= maxCharacters)
        {
            return trimmed;
        }

        // macOS reserves 36 characters for the marker before cutting.
        var keep = Math.Max(0, maxCharacters - 36);
        return trimmed[..keep].TrimEnd() + TruncationMarker;
    }

    /// <summary>
    /// A title for a tracker's single-line field: whitespace runs (including newlines) collapse to
    /// one space, the result is trimmed, and anything past <paramref name="maxCharacters"/> is cut
    /// with a single "…" — no "[truncated …]" suffix, which would consume most of a short field.
    /// </summary>
    public static string TrackerTitle(string? value, int maxCharacters)
    {
        var collapsed = string.Join(
            " ",
            (value ?? string.Empty).Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
        if (collapsed.Length <= maxCharacters)
        {
            return collapsed;
        }

        var keep = Math.Max(0, maxCharacters - 1);
        return collapsed[..keep].TrimEnd() + "…";
    }

    /// <summary>
    /// Truncates each entry and appends the macOS omission notice when entries had to be dropped.
    /// </summary>
    public static IReadOnlyList<string> LimitedList(
        IReadOnlyList<string> values,
        int maxItems,
        int maxCharactersPerItem)
    {
        var limited = values
            .Take(maxItems)
            .Select(value => Truncated(value, maxCharactersPerItem))
            .ToList();

        if (values.Count > maxItems)
        {
            limited.Add(OmissionNotice);
        }

        return limited;
    }
}

File: Sources/BugNarrator/Services/TrackerExportPayloadBudget.swift
import Foundation

enum TrackerExportPayloadBudget {
    static let gitHubBodyLimit = 16_000
    static let jiraTextLimit = 12_000
    static let issueSummaryLimit = 2_000
    static let evidenceLimit = 4_000
    static let noteLimit = 2_000
    static let metadataItemLimit = 12
    static let reproductionStepLimit = 10
    static let listEntryLimit = 500
    static let screenshotListLimit = 10
    /// Jira Cloud rejects a `summary` over 255 characters or containing a
    /// newline; GitHub rejects an issue `title` over 256. Both are hard server
    /// limits, unlike the body budgets above, which are self-imposed. Jira
    /// counts UTF-16 code units while `trackerTitle` counts Characters, so a
    /// title dense in astral-plane emoji or combining sequences can sit at the
    /// Character cap and still exceed Jira's — accepted for bug titles.
    static let jiraSummaryLimit = 255
    static let gitHubTitleLimit = 256

    static func truncated(_ value: String, maxCharacters: Int) -> String {
        let trimmedValue = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmedValue.count > maxCharacters else {
            return trimmedValue
        }

        let endIndex = trimmedValue.index(trimmedValue.startIndex, offsetBy: max(0, maxCharacters - 36))
        return trimmedValue[..<endIndex].trimmingCharacters(in: .whitespacesAndNewlines)
            + " …[truncated by BugNarrator for tracker limits]"
    }

    static func limitedList(
        _ values: [String],
        maxItems: Int,
        maxCharactersPerItem: Int
    ) -> [String] {
        let trimmedValues = values.prefix(maxItems).map {
            truncated($0, maxCharacters: maxCharactersPerItem)
        }

        if values.count > maxItems {
            return trimmedValues + ["Additional items were omitted by BugNarrator to fit tracker limits."]
        }

        return trimmedValues
    }

    /// A title for a tracker's single-line field: whitespace runs (including
    /// newlines) collapse to one space, the result is trimmed, and anything past
    /// `maxCharacters` is cut with a single "…" — no "[truncated …]" suffix,
    /// which would consume most of a short field.
    static func trackerTitle(_ value: String, maxCharacters: Int) -> String {
        let collapsed = value
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
        guard collapsed.count > maxCharacters else {
            return collapsed
        }

        let keep = max(0, maxCharacters - 1)
        let cut = collapsed[..<collapsed.index(collapsed.startIndex, offsetBy: keep)]
            .trimmingCharacters(in: .whitespaces)
        return cut + "…"
    }

    static func hardLimitMarkdown(_ value: String, maxCharacters: Int) -> String {
        truncated(value, maxCharacters: maxCharacters)
    }
}
